domain/autocert/state/state.go

```diff
@@ -0,0 +1,114 @@
+// Copyright 2023 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package state
+
+import (
+	"context"
+	"database/sql"
+
+	"github.com/canonical/sqlair"
+	"github.com/juju/errors"
+	"github.com/juju/utils/v3"
+
+	coreDB "github.com/juju/juju/core/database"
+	"github.com/juju/juju/domain"
+)
+
+// ExternalController represents a single row from the database when
+// external_controller is joined with external_controller_address and
+// external_model.
+type Autocert struct {
+	// Name is the autocert cache name (primary key).
+	Name string `db:"name"`
+
+	// Data represents the binary (encoded) contents of the autocert.
+	Data string `db:"data"`
+}
+
+type State struct {
+	*domain.StateBase
+}
+
+func NewState(factory coreDB.TxnRunnerFactory) *State {
+	return &State{
+		StateBase: domain.NewStateBase(factory),
+	}
+}
+
+// Put implements autocert.Cache.Put.
+func (st *State) Put(ctx context.Context, name string, data []byte) error {
+	db, err := st.DB()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	uuid, err := utils.NewUUID()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	q := `
+INSERT INTO autocert_cache (uuid, name, data, encoding)
+VALUES (?, ?, ?, 0)
+  ON CONFLICT(uuid) DO UPDATE SET data=excluded.data
```

Comment:
This should never happen. If we hit this it would be best to go out and buy a lottery ticket, as you've just struck gold.

In reality, we should always error out here, as you're potentially overwriting the data of some other uuid.

Reply:
:rofl: done